### DISCUSSION TITLE
Thinking to contribute to this project

### DISCUSSION THREAD

**Discussion by ShifatHossain** [hyOzd/serialplot#27] [General]
Hello,

I love the SerialPlot project as I use it all the time for my personal and professional work. And I have found some added features and some enhancements here and there that would be awesome for this project. I was considering to contribute this project by raising issues and proposing pull requests considering different issues raised. 

Do you have any brief guide about the source code organization or guide for development, that would help me to jump-start the development? 

Thanks.

**Comment by hyOzd** [comment]
@ShifatHossain Sorry for late reply. I was on a holiday.

SerialPlot isn't a big project. And there isn't much structure to it. Code is mostly designed around the GUI features. If you see a panel, there is probably a 'class' for it, and to work on that feature you should find that class and start from there. For example to add a new control to the serial port configuration you should work on 'PortControl' class. Data parsers are called readers, I recommend reading those. They aren't that complicated. Plot widget is based on 'Qwt' widget library. So if you want to make improvements on that, you should maybe check out Qwt first.

One thing I can add is the data flow. Where does the data come from and how it ends up displayed. I will try to be brief :)

Obviously data is read from 'serial port'. Most of the serialport handling is done by the Qt's serialport class. We only pass configuration parameters to that class. SerialPlot has 'reader' classes that read data from the serialport. For each data format (ascii, simple binary and custom format) there is a separate reader class. Selected **reader**, parses the incoming data and passes it to the `Stream` class. This is where data storage and channel configuration is managed. _Stream_ stores the actual data in `RingBuffer`s. It is called a ring buffer because we allocate a constant size of buffer one time and reuse it. As new data comes along, data is added to the end of it, while old data is removed from the beginning. Actual implementation is smart, memory isn't moved when adding new data. `Plot` widget, reads the data from the ring buffers for display. There isn't much to document about Plot widget because most of the job is done by the `QwtPlot`, but there are a few tricks that I had to do for optimization purposes. So it can be a bit difficult to do modifications on drawing side of things.

I hope this was a useful introduction. If you need help with a specific topic, please let me know.
